src/s_inter.c

```diff
@@ -201,30 +740,70 @@ static int sys_domicrosleep(int microsec, int pollem)
-                FD_SET(fp->fdp_fd, &readset);
-        if(select(pd_this->pd_inter->i_maxfd+1,
-                  &readset, &writeset, &exceptset, &timout) < 0)
-          perror("microsleep select");
+#ifdef THREADED_IO
+        // Pd was built with THREADED_IO support, but no one has taken
+        // charge of the IO duties, so we do it ourselves in the audio thread.
+        // Sub-optimal because of the extra memory copies, but better than
+        // nothing.
+        if(!pd_this->pd_inter->i_dontmanageio)
+            sys_doio(pd_this);
+        ring_buffer* rbrmfdsend = pd_this->pd_inter->i_rbrmfdsend;
+        if(rbrmfdsend)
+        {
+            // fds queued for deletion upon a failed send() call
+            while(rb_available_to_read(rbrmfdsend) >= sizeof(int))
+            {
+                int fd;
+                rb_read_from_buffer(rbrmfdsend, (char*)&fd, sizeof(fd));
+                printf("Removing send fd %d from send list\n", fd);
+                // find what object owns it
+                for(i = 0; i < pd_this->pd_inter->i_nfdsend; ++i)
+                {
+                    t_fdsend* fdsends = pd_this->pd_inter->i_fdsend;
+                    if(fd == fdsends[i].fds_fd)
+                    {
+                        printf("fd %d found to belong to %p, so calling %p\n", fd, fdsends[i].fds_ptr, fdsends[i].fds_fn);
+                        fdsends[i].fds_fn(fdsends[i].fds_ptr);
+                    }
+                }
+            }
+        }
+#endif // THREADED_IO
+        fd_set readset;
+        fdp_select(&readset, timout, kFdpManagerAudioThread);
         pd_this->pd_inter->i_fdschanged = 0;
         for (i = 0; i < pd_this->pd_inter->i_nfdpoll &&
             !pd_this->pd_inter->i_fdschanged; i++)
-                if (FD_ISSET(pd_this->pd_inter->i_fdpoll[i].fdp_fd, &readset))
         {
-            (*pd_this->pd_inter->i_fdpoll[i].fdp_fn)
-                (pd_this->pd_inter->i_fdpoll[i].fdp_ptr,
-                    pd_this->pd_inter->i_fdpoll[i].fdp_fd);
-            didsomething = 1;
+            t_fdpoll *fp = pd_this->pd_inter->i_fdpoll + i;
+            int fd = fp->fdp_fd;
+            int should_call = 0;
+            if(kFdpManagerAudioThread == fp->fdp_manager) {
+                if(FD_ISSET(fd, &readset)) {
+                    should_call = 1;
+                    didsomething = 1;
+                }
+            } else {
+#ifdef THREADED_IO
+                if(rbskt_ready(fp->fdp_rbskt))
+                    should_call = 1;
+#endif // THREADED_IO
+            }
+            if(should_call)
+            {
+                (*pd_this->pd_inter->i_fdpoll[i].fdp_fn)
+                    (pd_this->pd_inter->i_fdpoll[i].fdp_ptr,
+                        pd_this->pd_inter->i_fdpoll[i].fdp_fd);
+            }
+#ifdef THREADED_IO
+            // restore ownership to IO thread
+            if(kFdpManagerAudioThread == fp->fdp_manager)
+               fp->fdp_manager = kFdpManagerIoThread;
```

Comment:
First of all, is there a reason why the "ownership" is restored here and not right after https://github.com/pure-data/pure-data/pull/1261/files#diff-0a4510a7e0742e9142010737129f5bc23e6a31632899067dc8fe8e6deb47e389R785?

Also, `fdp_manager` is modified by different threads without synchronization and you have a classical ABA problem:
1) audio thread checks if `fdp_manager` is `kFdpManagerAudioThread`
2) I/O thread noticed that fd is ready and sets `fdp_manager` to `kFdpManagerAudioThread` to notify audio thread
3) audio thread sets `fdp_manager` to `kFdpManagerIoThread` and thus loses the notification.

In this case, it is not dangerous because the I/O thread will simply call select() again, but it is a bit sloppy.

The correct way is to perform an atomic store in the I/O thread and an atomic CAS in the audio thread. (If `fdp_manager` can only have two possible states, you can also use an atomic exchange.)

Reply:
> First of all, is there a reason why the "ownership" is restored here and not right after 

I don't think your link points to a specific line.

> 1. audio thread checks if fdp_manager is kFdpManagerAudioThread
> 2. I/O thread noticed that fd is ready and sets fdp_manager to kFdpManagerAudioThread to notify audio thread

I am not 100% sure, but I don't think this could happen: if `fdp_manager == kFdpManagerAudioThread` then the I/O thread will not call `select()` on the `fd` because it calls `fdp_select(&readset, timout, kFdpManagerIoThread);` (i.e.: it only selects on `fd`s that are managed by it). So, after the I/O thread has set `fdp_manager = kFdpManagerAudioThread`, it won't touch that `fd` again until control is transferred back to it from the audio thread with `fp->fdp_manager = kFdpManagerIoThread;`.

Reply:
> I don't think your link points to a specific line.

Sorry, I meant line 785: `if(kFdpManagerAudioThread == fp->fdp_manager)`


> I am not 100% sure, but I don't think this could happen

I think you're right, sorry for the noise. (But can you *prove* that the code is race free? ;-) Using atomics would at the very least make it obvious that the variable is shared between threads and used for synchronization. Also, I wouldn't have to think through all the possible scenarios of how the compiler (or the CPU) could reorder the code and break things.

Generally, you should use proper atomic operations to signal things between threads. If thread A does some task and wants to notify B on completion, it has to perform an atomic store (with release semantics) on the shared variable. When B does an atomic load (with aquire semantics) on the shared variable, it is guaranteed to see the finished task. Otherwise you might run into situations where in A the compiler moves the store *before* the end of the task and B would wrongly think it has completed. There is lots of old lockfree code out there which seems to work fine on x86, but breaks horribly on CPUs with weak memory models, like ARM.

Reply:
> First of all, is there a reason why the "ownership" is restored here and not right after
> ... line 785: if(kFdpManagerAudioThread == fp->fdp_manager)

I did it to prevent the IO thread from `select()`ing on it again and possibly set the flag again while the `fd` was read (possibly repeatedly) by the `pollfn` in the audio thread. Then the audio thread would read all the data from the socket, so that the flag would signal data available when there actually is no data available. One issue I encountered while working on this is that some of the `pollfn()`s fail if you call them without there being data available, they may fail and trigger a `sys_rmpollfn()` call, so this possibility has to be avoided. For this reason I am not transferring ownership back to the other thread until I am 100% done.

> (But can you prove that the code is race free? ;-)

Good question. On the IO thread side, writing to the ring buffer includes a memory barrier, so that should be fine. On the audio thread side, I figured that there is no way for the compiler to reorder writing to the global variable around the function pointer call. As the function content is not known at compile time, it cannot be inlined.

Anyhow, it seems that switching to atomics would give us peace of mind and clearer semantics, so I will do that. I think I will borrow the cross-platform macros defined in `ringbuffer.c`.

Reply:
> On the IO thread side, writing to the ring buffer includes a memory barrier, so that should be fine.

We don't necessarily write to the ring buffer.

> I figured that there is no way for the compiler to reorder writing to the global variable around the function pointer call.

That's true :-)

> Anyhow, it seems that switching to atomics would give us peace of mind and clearer semantics, so I will do that.

+1

Reply:
So if I understand correctly I actually need the following in each thread:
- atomic load (including memory barrier)
- do stuff
- atomic store (including memory barrier)
right?

Reply:
i.e.:

```
@@ -610,7 +636,7 @@ static void poll_fds()
             if(fp->fdp_callback_in_audio_thread)
             {
                 // temporarily transfer ownership
-                fp->fdp_manager = kFdpManagerAudioThread;
+                SYNC_STORE(&fp->fdp_manager, kFdpManagerAudioThread);
             }
             else
             {
@@ -745,7 +771,8 @@ static void fdp_select(fd_set *readset, struct timeval timout, const t_fdp_manag
     for (fp = pd_this->pd_inter->i_fdpoll,
         i = pd_this->pd_inter->i_nfdpoll; i--; fp++)
     {
-        if(kFdpManagerAnyThread == manager || manager == fp->fdp_manager)
+        t_fdp_manager fdp_manager = SYNC_FETCH(&(fp->fdp_manager));
+        if(kFdpManagerAnyThread == manager || manager == fdp_manager)
         {
             FD_SET(fp->fdp_fd, readset);
             ++count;
@@ -808,7 +835,8 @@ static int sys_domicrosleep(int microsec, int pollem)
             t_fdpoll *fp = pd_this->pd_inter->i_fdpoll + i;
             int fd = fp->fdp_fd;
             int should_call = 0;
-            if(kFdpManagerAudioThread == fp->fdp_manager) {
+            t_fdp_manager fdp_manager = SYNC_FETCH(fp->fdp_manager);
+            if(kFdpManagerAudioThread == fdp_manager) {
                 if(FD_ISSET(fd, &readset)) {
                     should_call = 1;
                     didsomething = 1;
@@ -828,7 +856,7 @@ static int sys_domicrosleep(int microsec, int pollem)
 #ifdef THREADED_IO
             // restore ownership to IO thread
             if(kFdpManagerAudioThread == fp->fdp_manager)
-               fp->fdp_manager = kFdpManagerIoThread;
+               SYNC_STORE(&fp->fdp_manager, kFdpManagerIoThread);
 #endif // THREADED_IO
         }
         if (didsomething)
```

(while I figure out how to do a cross-platform `SYNC_STORE()`)

Reply:
Here's an attempt  https://github.com/giuliomoro/pure-data-1/commit/2054c7970a1a6bcf085eb5cab22dcdeea5827f15
 
How does this look? My understanding at this point is that the atomic store and load are only an excuse to get the memory barrier that comes with them.

Reply:
>  So if I understand correctly I actually need the following in each thread:
>
> *  atomic load (including memory barrier)
> *  do stuff
> *  atomic store (including memory barrier)
>
> right?

You rarely need full memory barriers. Most of time you need aquire/release semantics (https://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering).

Read `fdp_manager` with `atomic_load_explicit(..., memory_order_acquire)` and write `fdp_manager` with `atomic_store_explicit(..., memory_order_release)`. I forgot what the platform specific versions are, but I can look it up... Lockfree programming in pre-C11 is a pain!

Note that in `sys_domicrosleep` the second load (after which you perform the store) can have `memory_order_relaxed` because the variable has been already aquired. (After all, it's possible to write the code entirely without the second if statement.)

> My understanding at this point is that the atomic store and load are only an excuse to get the memory barrier that comes with them. 

Atomic operations are *mainly* about synchronization. (On modern CPUs, read/write atomicity is not an actual issue for properly aligned variables up to 8 bytes.) Atomic doesn't necessarily imply memory barriers, either. On CPUs with strong memory ordering, like x86, aquire/release only prevents the compiler from reordering reads/stores. Only on CPUs with weak memory ordering, like ARM, we also need special CPU instructions or memory fences.

Note that `SYNC_FETCH` and `SYNC_COMPARE_EXCHANGE` (as defined in `ringbuffer.c`) are actually overkill for a lockfree ringbuffer. Both are read/modify/write operations, whereas we only need a simple atomic store and atomic load. Also, the loads in `rb_available_to_write` and `rb_available_to_read` can be relaxed, but that's a different story :-) Don't worry about that for now, I can take care of that after the PR gets merged.

Reply:
> (Note that while most of those macros come from ringbuffer.c, I added SYNC_STORE() myself, so it's worth double checking that.

The C11 version of `SYNC_STORE` is fine! The pre-C11 version (delegating to `SYNC_COMPARE_EXCHANGE` ) works, but is a bit overkill. But again, don't worry about that for now.

Reply:
And sorry for all the nitpicking. I just happened to write a lot of lockfree code lately, so I just wanted to give my 2 cents :-)

Reply:
> . After all, it's possible to write the code entirely without the second if statement - which IMO would be much clearer!

If I understand correctly, you are suggesting to do something like:

```
            t_fdp_manager fdp_manager = SYNC_FETCH(&fp->fdp_manager);
            if(kFdpManagerAudioThread == fdp_manager) {
                SYNC_STORE(&fp->fdp_manager, kFdpManagerIoThread); // <<< give it back immediately
                should_call = 1;
                didsomething = 1;
            }
            // call pollfn()
```
I think the SYNC_STORE() has to happen AFTER the call to `pollfn()` to ensure that the fd is "given back" to the IO thread only after all the operations on it are completed. Otherwise, the IO thread could mark it as "ready" while the audio thread is still reading from it. 

> the second load (after which you perform the store) can have memory_order_relaxed because the variable has been already aquired.

yes, in my current implementation I am just reading it without any synchronisation, because it's already been acquired.

> And sorry for all the nitpicking. 

Thanks for doing that, I'd love to learn more about these things, but I never had to deal with them (and always just copied `ringbuffer.c/h` into whatever project needed lock-free buffers!). So right now I don't think I can do any better than what I have already done in 21ad0c7356bd1938f0beda7ba54461d51aeb2ed6 ... 
If that's good enough for now, then I will just wait for ...

> Don't worry about that for now, I can take care of that after the PR gets merged.

Reply:
> If I understand correctly, you are suggesting to do something like:

No, I meant something like this:

```
#ifdef THREADED_IO
            t_fdp_manager fdp_manager = SYNC_FETCH(&fp->fdp_manager);
            if (kFdpManagerAudioThread == fdp_manager) {
                (*pd_this->pd_inter->i_fdpoll[i].fdp_fn)
                    (pd_this->pd_inter->i_fdpoll[i].fdp_ptr,
                        pd_this->pd_inter->i_fdpoll[i].fdp_fd);
                didsomething = 1;
                // restore ownership to IO thread
                SYNC_STORE(&fp->fdp_manager, kFdpManagerIoThread);
            } else {
                if (rbskt_ready(fp->fdp_rbskt))
                    (*pd_this->pd_inter->i_fdpoll[i].fdp_fn)
                        (pd_this->pd_inter->i_fdpoll[i].fdp_ptr,
                            pd_this->pd_inter->i_fdpoll[i].fdp_fd);
            }
#else // THREADED_IO
            if (kFdpManagerAudioThread == fp->fdp_manager) {
                if (FD_ISSET(fd, &readset)) {
                    (*pd_this->pd_inter->i_fdpoll[i].fdp_fn)
                        (pd_this->pd_inter->i_fdpoll[i].fdp_ptr,
                            pd_this->pd_inter->i_fdpoll[i].fdp_fd);
                    didsomething = 1;
                }
            }
#endif // THREADED_IO
```
I guess you tried to avoid duplicating the monstrous function pointer call. Instead of complicating the logic, you could just add a helper function:
```
static void callpollfn(int i)
{
    (*pd_this->pd_inter->i_fdpoll[i].fdp_fn)
        (pd_this->pd_inter->i_fdpoll[i].fdp_ptr,
            pd_this->pd_inter->i_fdpoll[i].fdp_fd);
}

```
So the above becomes:
```
#ifdef THREADED_IO
            t_fdp_manager fdp_manager = SYNC_FETCH(&fp->fdp_manager);
            if (kFdpManagerAudioThread == fdp_manager) {
                callpollfn(i);
                didsomething = 1;
                // restore ownership to IO thread
                SYNC_STORE(&fp->fdp_manager, kFdpManagerIoThread);
            } else {
                if (rbskt_ready(fp->fdp_rbskt))
                    callpollfn(i);
            }
#else // THREADED_IO
            if (kFdpManagerAudioThread == fp->fdp_manager) {
                if (FD_ISSET(fd, &readset)) {
                    callpollfn(i);
                    didsomething = 1;
                }
            }
#endif // THREADED_IO
```
IMHO, this is both shorter and easier to follow.

> I think the SYNC_STORE() has to happen AFTER the call to pollfn() to ensure that the fd is "given back" to the IO thread only after all the operations on it are completed. 

Yes, of course!

Reply:
Oh OK. Yes, I was trying to avoid to duplicate the ugly call. Here's the commit that addresses this:

142f0375e380babf04702d4df9cc05dd73fca25b